master/buildbot/db/connector.py

```diff
@@ -148,6 +149,8 @@ def setup(self, check_version=True, verbose=True):
         self._engine = enginestrategy.create_engine(db_url, basedir=self.basedir)
         self.upsert = get_upsert_method(self._engine)
         self.pool = pool.DBThreadPool(self._engine, reactor=self.master.reactor, verbose=verbose)
+        self._db_tasks_waiter = DeferWaiter()
+        self.master.reactor.addSystemEventTrigger('before', 'shutdown', self._db_tasks_waiter.wait)
```

Comment:
I think `stopService()` is enough. It is a bug to not wait for `stopService()` to finish before shutting down the reactor.

Reply:
I'm not particularly happy with this.
Lifetime of services are pretty confusing and all over the place.

`DBThreadPool` (and compression pool now) is stopped on the `during, shutdown` event of the reactor.

But this is "stopped" on `DBConnector.stopService`.
I also had to make a `_shutdown` method since, in tests, services are not started nor then stopped.

I feel it'd be more consistent to either:
- `self._db_tasks_waiter.wait()` on `before, shutdown` of the reactor
- add start/stop methods to `DBConnectorComponent` / `DBThreadPool` and call them from the `DBConnector` `startService` / `stopService`.